File: Java/RecyclerView9RecyclerViewAsViewPager/app/src/main/java/dev/ogabek/recyclerview9_recyclerviewasviewpager/activity/MainActivity.java
package dev.ogabek.recyclerview9_recyclerviewasviewpager.activity;

import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.PagerSnapHelper;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SnapHelper;

import android.content.Context;
import android.os.Bundle;
import android.util.Log;

import java.util.ArrayList;
import java.util.List;

import dev.ogabek.recyclerview9_recyclerviewasviewpager.R;
import dev.ogabek.recyclerview9_recyclerviewasviewpager.adapter.MainRecyclerViewAdapter;
import dev.ogabek.recyclerview9_recyclerviewasviewpager.model.Member;

public class MainActivity extends AppCompatActivity {

    Context context;
    RecyclerView recyclerView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        initViews();

        refreshAdapter(prepareMemberList());

    }

    private void refreshAdapter(List<Member> members) {
        MainRecyclerViewAdapter adapter = new MainRecyclerViewAdapter(context, members);
        recyclerView.setAdapter(adapter);
    }

    private List<Member> prepareMemberList() {
        List<Member> members = new ArrayList<>();
        for (int i = 0; i < 5; i++) {
            members.add(new Member("Ogabek", i));
        }
        return members;
    }

    private void initViews() {
        context = this;
        recyclerView = findViewById(R.id.rv_mainRecyclerView);
        recyclerView.setLayoutManager(new LinearLayoutManager(context, LinearLayoutManager.HORIZONTAL, false));
        SnapHelper snapHelper = new PagerSnapHelper();
        snapHelper.attachToRecyclerView(recyclerView);

        recyclerView.setOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrollStateChanged(@NonNull RecyclerView recyclerView, int newState) {
                super.onScrollStateChanged(recyclerView, newState);
                LinearLayoutManager manager = (LinearLayoutManager) recyclerView.getLayoutManager();
                int activePosition = 0;
                if (manager != null) {
                    activePosition = manager.findFirstVisibleItemPosition();
                }
                if (activePosition != RecyclerView.NO_POSITION) return;
                Log.d("Active Position", "Active Position: " + activePosition);
            }

            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                super.onScrolled(recyclerView, dx, dy);
            }
        });
    }

}

File: Kotlin/RecyclerView9RecyclerViewAsViewPager/app/src/main/java/dev/ogabek/recyclerview_9_recyclerviewasviewpager/activity/MainActivity.kt
package dev.ogabek.recyclerview_9_recyclerviewasviewpager.activity

import android.content.Context
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.PagerSnapHelper
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.SnapHelper
import dev.ogabek.recyclerview_9_recyclerviewasviewpager.R
import dev.ogabek.recyclerview_9_recyclerviewasviewpager.adapter.MainAdapter
import dev.ogabek.recyclerview_9_recyclerviewasviewpager.model.Ads
import java.util.ArrayList

class MainActivity : AppCompatActivity() {
    private lateinit var context: Context
    private lateinit var recyclerView: RecyclerView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        initViews()

        refreshAdapter(prepareMemberList())

    }

    private fun initViews() {
        context = this
        recyclerView = findViewById(R.id.rv_main_recycler_view)
        recyclerView.layoutManager = LinearLayoutManager(context, LinearLayoutManager.VERTICAL, false)
        val snapHelper: SnapHelper = PagerSnapHelper()
        snapHelper.attachToRecyclerView(recyclerView)

        recyclerView.setOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                super.onScrollStateChanged(recyclerView, newState)
                val manager = recyclerView.layoutManager as LinearLayoutManager?
                var activePosition = 0
                if (manager != null) {
                    activePosition = manager.findFirstVisibleItemPosition()
                }
                if (activePosition != RecyclerView.NO_POSITION) return
                Log.d("Active Position", "Active Position: $activePosition")
            }
        })
    }

    private fun refreshAdapter(ads: List<Ads>) {
        val adapter = MainAdapter(context, ads)
        recyclerView.adapter = adapter
    }

    private fun prepareMemberList(): List<Ads> {
        val members: MutableList<Ads> = ArrayList<Ads>()
        for (i in 0..50) {
            members.add(Ads())
        }
        return members
    }
}